### DISCUSSION TITLE
Component template counts increase when no component is effectively added

### DISCUSSION THREAD

**Discussion by stanislav-zaprudskiy** [netbox-community/netbox#16415] [General]
It's a re-post of [this Slack thread](https://netdev-community.slack.com/archives/C01P0FRSXRV/p1717517761101679), which didn't get any response.

I'm observing some unexpected behavior when dealing with component template counts on DeviceType (https://github.com/netbox-community/netbox/issues/13269 / https://github.com/netbox-community/netbox/issues/6347). Here's the test:

```
    def test_relation_add(self):
        """
        Counters are updated on idempotent relation `add()`
        """
        manufacturer, _ = Manufacturer.objects.get_or_create(name='Manufacturer 1', slug='manufacturer-1')
        devicetype = DeviceType.objects.create(model='Device Type X', slug='dt-x', manufacturer=manufacturer)

        powerporttemplate = PowerPortTemplate()
        powerporttemplate.device_type = devicetype
        powerporttemplate.name = "ps-1"
        powerporttemplate.save()

        # https://docs.djangoproject.com/en/4.2/ref/models/relations/#django.db.models.fields.related.RelatedManager.add:
        # > Using add() on a relation that already exists won't duplicate the
        # > relation, but it will still trigger signals.
        devicetype.powerporttemplates.add(powerporttemplate, bulk=False)
        devicetype.refresh_from_db()
        self.assertEqual(devicetype.power_port_template_count, 1)
```

which fails with:

```
FAIL: test_relation_add (utilities.tests.test_counters.CountersTest.test_relation_add)
Counters are updated on idempotent relation `add()`
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/netbox/netbox/utilities/tests/test_counters.py", line 106, in test_relation_add
    self.assertEqual(devicetype.power_port_template_count, 1)
AssertionError: 2 != 1
```

So the count got increased, even though `add()` didn't add the `powerporttemplate` (as it's been already associated).

I believe it updates the counter because [the corresponding `instance.tracker`](https://github.com/netbox-community/netbox/blob/87109f5539758a3d62ea8200d89641b21bb2d283/netbox/utilities/counters.py#L55) doesn't get cleared [upon `.save()`](https://github.com/netbox-community/netbox/blob/87109f5539758a3d62ea8200d89641b21bb2d283/netbox/utilities/tracking.py#L55), thus it wrongly assumes that it needs to update it. Am I doing something conceptually wrong in my code, or is it Netbox which would benefit from a fix? Thanks!

**Comment by rizlas** [comment]
Same issue here https://github.com/netbox-community/netbox/discussions/15718 and https://github.com/netbox-community/netbox/discussions/17900

Check this https://github.com/netbox-community/customizations/issues/100#issuecomment-2026022272
